aten/src/ATen/xpu/CachingHostAllocator.cpp

```diff
@@ -0,0 +1,77 @@
+#include <ATen/xpu/CachingHostAllocator.h>
+
+namespace at::xpu {
+namespace {
+
+constexpr size_t kHostAlignment = 512;
+
+using Block = HostBlock<XPUStream>;
+using AllocatorImplInterface =
+    CachingHostAllocatorImplInterface<XPUStream, XPUEvent>;
+
+struct XPUHostAllocatorImpl : public AllocatorImplInterface {
+  /* These following functions are runtime-related. */
+  void allocate_host_memory(size_t size, void** ptr) override {
+    *ptr = sycl::aligned_alloc_host(
+        kHostAlignment, size, c10::xpu::get_device_context());
+  }
+
+  void free_block(Block* block) override {
+    sycl::free(block->ptr_, c10::xpu::get_device_context());
+  }
+
+  void record_stream(
+      c10::optional<std::vector<XPUEvent>>& events,
+      XPUStream stream) override {
+    XPUEvent event;
+    event.record(stream);
+    events->push_back(std::move(event));
+  }
+
+  bool query_event(XPUEvent& event) override {
+    return event.query();
+  }
+};
+
+} // anonymous namespace
+
+void raw_local_deleter(void* ptr);
+
+struct XPUHostAllocator final
+    : public HostAllocatorInterface<XPUHostAllocatorImpl> {
+  at::DataPtr allocate(size_t size) override {
+    auto ptr_and_ctx = impl_->allocate(size);
+    return {
+        ptr_and_ctx.first,
+        ptr_and_ctx.second,
+        &raw_local_deleter,
+        at::DeviceType::CPU};
+  }
+};
+
+static XPUHostAllocator host_allocator;
```

Comment:
Should we move `host_allocator` to the anonymous namespace?

Reply:
It is ok to move to an anonymous namespace. Done.